clash-prelude/src/Clash/Class/Finite/Internal.hs

```diff
@@ -0,0 +1,1006 @@
+import Clash.Sized.Internal.Unsigned (Unsigned(..))
+import Clash.Sized.Signed (Signed)
+import Clash.Sized.RTree (RTree(..), tdfold, tfold, trepeat)
+import Clash.Sized.Vector
+  ( Vec(..), (++), (!!), concatMap, bv2v, dfold, foldl, foldr
+  , ifoldr, indicesI, iterateI, unfoldrI, repeat, reverse, replace
+  )
+import Clash.XException (ShowX, NFDataX, errorX)
+
+import qualified Data.List as List (iterate)
+
+{- $setup
+>>> :m -Prelude
+>>> :set -XDeriveAnyClass
+>>> import Clash.Prelude
+>>> import Data.Ord
+-}
+
+-- * Finite Class
+
+-- | The class of types holding only a finite number of elements.
+--
+-- The class supports generic deriving, i.e., for custom types the
+-- class instances can be derived via @derive (Generic, Finite)@
+-- requiring that all inner types of the type declaration have
+-- @Finite@ instances as well.
+--
+-- >>> data T = B Bit | D (Index 2) (Signed 1) deriving (Generic, Finite, Show)
+-- >>> natToNum @(ElementCount T)
+-- 6
+-- >>> elements @T
+-- B 0 :> B 1 :> D 0 -1 :> D 0 0 :> D 1 -1 :> D 1 0 :> Nil
+-- >>> lowestMaybe @T
+-- Just (B 0)
+-- >>> highestMaybe @T
+-- Just (D 1 0)
+-- >>> succMaybe (B 1)
+-- Just (D 0 -1)
+-- >>> predMaybe (B 0)
+-- Nothing
+-- >>> index (D 0 0)
+-- 3
+-- >>> ith @T 5
+-- D 1 0
+--
+-- Any definition must satisfy the following laws (automatically
+-- ensured when generic deriving the instance):
+--
+-- [Index Order]
+--   @ index '<$>' elements = 'indicesI' @
+-- [Forward Iterate]
+--   @ 'iterateI' ('>>=' succMaybe) (lowestMaybe \@a)
+--      = 'Just' '<$>' (elements \@a) @
+-- [Backward Iterate]
+--   @ 'iterateI' ('>>=' predMaybe) (highestMaybe \@a)
+--      = 'Just' '<$>' 'reverse' (elements \@a) @
+-- [Index Isomorphism]
+--   @ith (index x) = x@
+-- [Minimum Predecessor]
+--   @ lowestMaybe '>>=' predMaybe = 'Nothing' @
+-- [Maximum Successor]
+--   @ highestMaybe '>>=' succMaybe = 'Nothing' @
+-- [Extremes]
+--   @ ElementCount a = 0:
+--       lowestMaybe \@a  = 'Nothing',
+--       highestMaybe \@a = 'Nothing' @
+--  [ ] @ ElementCount a > 0:
+--       lowestMaybe \@a  = 'Just' lowest,
+--       highestMaybe \@a = 'Just' highest @
+--
+-- If @a@  has a 'Bounded' instance, it further must satisfy:
+--
+-- [Bounded Compatibility]
+--   @ ElementCount a > 0: lowest \@a
+--      = 'minBound', highest \@a = 'maxBound' @
+--
+-- If @a@ has an 'Enum' instance, it further must satisfy:
+--
+-- [Enum Compatibility]
+--   @ succMaybe x = Just y: 'succ' x = y @
+--  [ ] @ predMaybe x = Just y: 'pred' x = y @
+--
+class KnownNat (ElementCount a) => Finite a where
+  -- | The number of elements of the type.
+  type ElementCount a :: Nat
+  type ElementCount a = GElementCount (Rep a)
+
+  -- | The elements of the type.
+  elements :: Vec (ElementCount a) a
+  default elements ::
+    ( Generic a, GFinite (Rep a)
+    , ElementCount a ~ GElementCount (Rep a)
+    ) => Vec (ElementCount a) a
+  elements = to <$> gElements
+
+  -- | The @0@ indexed element.
```

Comment:
```suggestion
  -- | The element at index 0.
```

Just a suggestion... it's just that to me it feels a zero-indexed something is something that has indexes starting at 0, like most arrays and stuff in computer science.